main/src/ca/uwaterloo/flix/language/ast/Ast.scala

```diff
@@ -374,7 +374,75 @@ object Ast {
   /**
     * Represents a label in the labelled graph.
     */
-  case class Label(pred: Name.Pred, arity: Int)
+  case class Label(pred: Name.Pred, den: Ast.Denotation, terms: List[Type], arity: Int)
+
+  /**
+    * Compare the labels by field, but not exactly for types.
+    * This will always correctly return true but might deem different labels
+    * equal depending on their term types.
+    */
+  def labelEquality(l1: Label, l2: Label): Boolean =
+    l1.pred == l2.pred &&
+      l1.den == l2.den &&
+      l1.terms.zip(l2.terms).forall { case (t1, t2) => typeConservativeEquality(t1, t2) } &&
+      l1.arity == l2.arity
+
+  /**
+    * Returns true if the types are equal but may return anything for different types.
+    */
+  private def typeConservativeEquality(tpe1: Type, tpe2: Type): Boolean = {
```

Comment:
OK. You have convinced me we should just use unification. You can call:

https://github.com/flix/flix/blob/master/main/src/ca/uwaterloo/flix/language/phase/unification/Unification.scala#L440

(I am still unsure if this has negative performance implications- but we have a research question for that).